File: java/src/main/java/com/warehouse/adapters/InMemoryInventoryGateway.java
package com.warehouse.adapters;

import com.warehouse.application.InventoryGateway;
import com.warehouse.domain.DomainError.CompensationFailureError;
import com.warehouse.domain.DomainError.InsufficientStockError;
import com.warehouse.domain.OrderId;
import com.warehouse.domain.OrderLine;
import com.warehouse.domain.Sku;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

/**
 * {@link InventoryGateway} double enforcing a finite stock map, for tests and demos.
 *
 * <p>{@code reserveAll} is atomic: every line is covered or none are. Concurrent callers serialize
 * on an internal lock so two reservations cannot oversell the last units.
 */
public final class InMemoryInventoryGateway implements InventoryGateway {

  private record Held(Sku sku, int amount) {}

  private final Lock lock = new ReentrantLock();
  private final Map<Sku, Integer> stock;
  private final Map<String, List<Held>> reservations = new HashMap<>();
  private boolean failRelease;

  /**
   * Starts from an initial per-SKU stock map, copied defensively.
   *
   * @param initialStock available units per SKU; missing SKUs mean zero stock
   */
  public InMemoryInventoryGateway(Map<Sku, Integer> initialStock) {
    this.stock = new HashMap<>(initialStock);
  }

  /**
   * Configures whether subsequent {@link #release} calls fail compensation.
   *
   * @param fail when true, {@link #release} returns {@link ReleaseOutcome.Failed}
   */
  public void failRelease(boolean fail) {
    lock.lock();
    try {
      this.failRelease = fail;
    } finally {
      lock.unlock();
    }
  }

  /**
   * Returns a copy of remaining units per SKU.
   *
   * @return remaining stock, independent of internal state
   */
  public Map<Sku, Integer> snapshotStock() {
    lock.lock();
    try {
      return Map.copyOf(stock);
    } finally {
      lock.unlock();
    }
  }

  @Override
  public ReservationOutcome reserveAll(
      OrderId orderId, List<OrderLine> lines, String idempotencyKey) {
    lock.lock();
    try {
      if (reservations.containsKey(idempotencyKey)) {
        return new ReservationOutcome.Reserved(new ReservationToken(orderId, idempotencyKey));
      }
      for (OrderLine line : lines) {
        int available = stock.getOrDefault(line.sku(), 0);
        if (available < line.quantity().value()) {
          return new ReservationOutcome.Shortage(
              new InsufficientStockError(line.sku(), line.quantity(), available));
        }
      }
      List<Held> held = new ArrayList<>();
      for (OrderLine line : lines) {
        int available = stock.getOrDefault(line.sku(), 0);
        stock.put(line.sku(), available - line.quantity().value());
        held.add(new Held(line.sku(), line.quantity().value()));
      }
      reservations.put(idempotencyKey, held);
      return new ReservationOutcome.Reserved(new ReservationToken(orderId, idempotencyKey));
    } finally {
      lock.unlock();
    }
  }

  @Override
  public ReleaseOutcome release(ReservationToken token) {
    lock.lock();
    try {
      if (failRelease) {
        return new ReleaseOutcome.Failed(new CompensationFailureError("release", "forced failure"));
      }
      List<Held> held = reservations.remove(token.idempotencyKey());
      if (held == null) {
        return new ReleaseOutcome.Released();
      }
      for (Held item : held) {
        stock.put(item.sku(), stock.getOrDefault(item.sku(), 0) + item.amount());
      }
      return new ReleaseOutcome.Released();
    } finally {
      lock.unlock();
    }
  }
}

File: kotlin/src/main/kotlin/com/warehouse/adapters/InMemoryInventoryGateway.kt
package com.warehouse.adapters

import com.warehouse.application.InventoryGateway
import com.warehouse.application.InventoryGateway.ReleaseOutcome
import com.warehouse.application.InventoryGateway.ReservationOutcome
import com.warehouse.application.InventoryGateway.ReservationToken
import com.warehouse.domain.CompensationFailureError
import com.warehouse.domain.InsufficientStockError
import com.warehouse.domain.OrderId
import com.warehouse.domain.OrderLine
import com.warehouse.domain.Sku
import java.util.concurrent.locks.ReentrantLock
import kotlin.concurrent.withLock

/**
 * [InventoryGateway] double enforcing a finite stock map, for tests and demos.
 *
 * `reserveAll` is atomic: every line is covered or none are. Concurrent callers
 * serialize on an internal lock so two reservations cannot oversell the last units.
 */
class InMemoryInventoryGateway(
    initialStock: Map<Sku, Int> = emptyMap()
) : InventoryGateway {
    private data class Held(
        val sku: Sku,
        val amount: Int
    )

    private val lock = ReentrantLock()
    private val stock = HashMap(initialStock)
    private val reservations = HashMap<String, List<Held>>()
    private var failRelease: Boolean = false

    fun failRelease(fail: Boolean) {
        lock.withLock { failRelease = fail }
    }

    fun snapshotStock(): Map<Sku, Int> = lock.withLock { stock.toMap() }

    override fun reserveAll(
        orderId: OrderId,
        lines: List<OrderLine>,
        idempotencyKey: String
    ): ReservationOutcome {
        return lock.withLock {
            if (idempotencyKey in reservations) {
                return@withLock ReservationOutcome.Reserved(
                    ReservationToken(orderId, idempotencyKey)
                )
            }
            for (line in lines) {
                val available = stock[line.sku] ?: 0
                if (available < line.quantity.value) {
                    return@withLock ReservationOutcome.Shortage(
                        InsufficientStockError(line.sku, line.quantity, available)
                    )
                }
            }
            val held = ArrayList<Held>(lines.size)
            for (line in lines) {
                val available = stock[line.sku] ?: 0
                stock[line.sku] = available - line.quantity.value
                held.add(Held(line.sku, line.quantity.value))
            }
            reservations[idempotencyKey] = held
            ReservationOutcome.Reserved(ReservationToken(orderId, idempotencyKey))
        }
    }

    override fun release(token: ReservationToken): ReleaseOutcome {
        return lock.withLock {
            if (failRelease) {
                return@withLock ReleaseOutcome.Failed(
                    CompensationFailureError("release", "forced failure")
                )
            }
            val held = reservations.remove(token.idempotencyKey)
            if (held != null) {
                for (item in held) {
                    stock[item.sku] = (stock[item.sku] ?: 0) + item.amount
                }
            }
            ReleaseOutcome.Released
        }
    }
}
